File: ruby/blockchain.rb
require 'digest/sha2'
require 'json'
require 'uri'
require 'faraday'

class Blockchain
  attr_reader :chain, :nodes

  def initialize
    @chain = []
    @current_transactions = []
    @nodes = []

    new_block(previous_hash: 1, proof: 100)
  end

  def register_node(address)
    uri = URI.parse(address)

    @nodes << "#{uri.scheme}://#{uri.host}:#{uri.port}"
    @nodes.uniq!
  end

  def valid_chain?(chain)
    last_block = chain[0]

    (1...chain.size).each do |index|
      block = chain[index]
      puts block

      return false unless block[:previous_hash] == self.class.hash(last_block)
      return false unless self.class.valid_proof?(last_block[:proof], block[:proof])

      last_block = block
    end

    true
  end

  def resolve_conflicts
    neighbours = @nodes
    new_chain = nil

    max_length = @chain.size
    neighbours.each do |node|
      response = Faraday.get("#{node}/chain")

      next unless response.status == 200

      json = JSON.parse(response.body, symbolize_names: true)

      length = json[:length]
      chain = json[:chain]

      if length > max_length && valid_chain?(chain)
        max_length = length
        new_chain = chain
      end
    end

    return false unless new_chain

    @chain = new_chain
    true
  end

  def new_block(previous_hash: nil, proof: nil)
    block = {
      index: @chain.size + 1,
      timestamp: Time.now,
      transactions: @current_transactions,
      proof: proof,
      previous_hash: previous_hash || self.class.hash(@chain[-1])
    }

    @current_transactions = []
    @chain << block

    block
  end

  def new_transaction(sender, recipient, amount)
    @current_transactions << {
      sender: sender,
      recipient: recipient,
      amount: amount
    }

    last_block[:index] + 1
  end

  def proof_of_work(last_proof)
    proof = 0
    while not self.class.valid_proof?(last_proof, proof)
      proof += 1
    end
    
    proof
  end

  def self.hash(block)
    Digest::SHA256.hexdigest block.to_json
  end

  def self.valid_proof?(last_proof, proof)
    guess = "#{last_proof}#{proof}"

    Digest::SHA256.hexdigest(guess)[0..3] == '0000'
  end

  def last_block
    @chain[-1]
  end
end

File: src/main/kotlin/Blockchain.kt
import com.sun.jmx.snmp.Timestamp
import java.security.MessageDigest

object Blockchain {
    val chain: MutableList<Block>? = null
    val currentTransactions: MutableList<Transaction>? = null

    init {
        val genesisBlock: Block = Block(
                1,
                System.currentTimeMillis()/1000,
                currentTransactions,
                100,
                "1"
        )
        // ジェネシスブロック追加
        newBlock(genesisBlock)
    }

    fun newBlock(block: Block){
        // 登録されていたtransactionをリセット
        currentTransactions?.dropWhile { true }
        // 新しいブロックを登録
        chain?.add(block)
    }

    fun newTransaction(transaction: Transaction) {
        // 新しいトランザクションをリストに加える
        currentTransactions?.add(transaction)
    }

    fun hash(block: Block): String {
        return convertHash(block.toString())
    }

    // hash(前のproof + proof)の最初の4文字が0となるhash値を探す
    fun proofOfWork(lastProof: String): String {
        var proof = 0
        while (!validWork(lastProof, Integer.toString(proof))) {
            proof += 1
        }
        return Integer.toString(proof)
    }

    fun validWork(lastProof: String, proof: String): Boolean {
        val hashVal = convertHash(lastProof + proof)
        return hashVal.substring(0, 4) == "0000"
    }

    fun convertHash(string: String): String {
        val digest = MessageDigest.getInstance("SHA-256").digest(string.toByteArray())
        val hashVal = digest.fold("", {str, it -> str + "%02x".format(it)})

        return hashVal
    }

}